src/nvim/eval.c

```diff
@@ -9828,34 +9829,15 @@ static void f_getcwd(typval_T *argvars, typval_T *rettv)
     scope_number[i] = argvars[i].vval.v_number;
     // The scope is the current iteration step.
     scope = i;
-
-    if (scope_number[i] < -1) {
-      EMSG(_(e_invarg));
-      return;
-    }
-  }
-
-  // Allocate and initialize the string to return.
-  cwd = xmalloc(MAXPATHL);
-
-  // Get the scope and numbers from the arguments
-  for (int i = 0; i < MAX_CD_SCOPE; i++) {
-    // If there is no argument there are no more scopes after it, break out.
-    if (argvars[i].v_type == VAR_UNKNOWN) {
-      break;
-    }
-    scope_number[i] = argvars[i].vval.v_number;
-    // The scope is the current iteration step.
-    scope = i;
     // It is an error for the scope number to be less than `-1`.
     if (scope_number[i] < -1) {
       EMSG(_(e_invarg));
-      goto end;
+      return;
     }
   }
 
-  // If the deepest scope number is `-1` advance the scope.
-  if (scope_number[scope] < 0) {
+  // If a scope number is `-1` advance the scope by one.
+  while (scope_number[scope] < 0 && scope < MAX_CD_SCOPE) {
     scope++;
```

Comment:
Another idea would be to forbid the last argument from being `-1` instead of implicitly moving a scope up. So `getcwd(-1)` throws an error instead of being equivalent to `getcwd(-1,0)`. Similarly `getcwd(-1,-1)` would be an error as well. But what about is we want the global working directory? In that case the user would have to call `getcwd(-1,-1,0)`, but this would break if a scope was later added between tab and global because the `0` would no longer refer to global but to whatever comes in between. I don't think we should do that.

You right about making the comment less verbose, I'll split it up in two comments, one before the `if` and one inside, that's still the same amount of extra lines, but it's easier on the eyes.